pkg/datastore/pagination/iterator.go

```diff
@@ -0,0 +1,114 @@
+package pagination
+
+import (
+	"context"
+	"errors"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// NewPaginatedIterator creates an implementation of the datastore.Iterator
+// interface that internally paginates over datastore results.
+func NewPaginatedIterator(
+	ctx context.Context,
+	reader datastore.Reader,
+	filter datastore.RelationshipsFilter,
+	pageSize uint64,
+	order options.SortOrder,
+) (datastore.RelationshipIterator, error) {
+	pi := &paginatedIterator{
+		ctx:      ctx,
+		reader:   reader,
+		filter:   filter,
+		pageSize: pageSize,
+		order:    order,
+		delegate: common.NewSliceRelationshipIterator(nil, options.ByResource),
+	}
+
+	pi.startNewBatch(nil)
+
+	return pi, pi.err
+}
+
+type paginatedIterator struct {
+	ctx      context.Context
+	reader   datastore.Reader
+	filter   datastore.RelationshipsFilter
+	pageSize uint64
+	order    options.SortOrder
+
+	delegate          datastore.RelationshipIterator
+	returnedFromBatch uint64
+	err               error
+	closed            bool
+}
+
+func (pi *paginatedIterator) Next() *core.RelationTuple {
+	if pi.Err() != nil {
+		return nil
+	}
+
+	var next *core.RelationTuple
+	for next = pi.delegate.Next(); next == nil; next = pi.delegate.Next() {
+		if pi.delegate.Err() != nil {
+			pi.err = pi.delegate.Err()
+			return nil
+		}
+
+		if pi.returnedFromBatch < pi.pageSize {
+			// No more tuples to get
+			return nil
+		}
+
+		cursor, err := pi.delegate.Cursor()
+		if errors.Is(err, datastore.ErrCursorEmpty) {
```

Comment:
it would seem we could ignore the error here if it's not `ErrCursorEmpty`

Reply:
Good catch!